### ISSUE TITLE
Support for arbitrary PersistentVolumes as backing storage

### ISSUE THREAD

**Issue by travisn** [rook/rook#919]
To run more smoothly in various cloud providers, the OSDs should be able to run on top of arbitrary volumes. Persisted config folders for mons and osds should also use the same volumes, which is related to #796.

**Comment by bassam** [comment]
Not sure I follow. Is this not a local storage scenario #796 ?

**Comment by travisn** [comment]
@bassam as i recall from a discussion we had a while back, the thought was that instead of hostPath or local storage under the OSD, the OSD would be able to mount any type of PV. But I don't recall why that was interesting since you could just mount the device from my previous comment, so I'll close this issue unless you recall something else.

**Comment by bassam** [comment]
Sorry not being clear. I think the OSDs should use any PV including local storage PVs. If someone wants to run OSDs onto of EBS, for example, they can. If they want to use instance storage then they can just use local storage PVs. This issue and #796 seem to cover the same work, no?

**Comment by travisn** [comment]
I see #796 specifically for local storage. If we're going to support other PVs it seems, each type would have its own parameters that would need to be exposed in some way in the CRD. Wouldn't that be a work item for each type of PV to support? What would be the scenario to support this? Rook's strength is that it can use local storage. And if you did need other volumes, could you mount them in a place that Rook could find them under `/dev`?

**Comment by bassam** [comment]
> What would be the scenario to support this?
I've heard from a number of users that they would feel better about running onto of EBS even initially. Other cloud-native storage systems support this too.

> And if you did need other volumes, could you mount them in a place that Rook could find them under /dev
I think we should avoid any operations that assume paths on the host or really anything about the host.

**Comment by travisn** [comment]
Based on slack conversation with @StianOvrevage and @jbw976 , the design might look like the following:
- Volume claims can be specified in the cluster CRD under the `storage` element.
- Claims will be a list under a `volumeClaims` element, and would be a sibling of the `nodes` and `directories` elements.
- The operator will create one osd per claim. The osd will assume usage of the entire volume.
- The operator will not assign the claim to a node. Kubernetes will handle that under the covers. This will be natural for local storage claims as well, where kubernetes will assign affinity to a node.

**Comment by jbw976** [comment]
This has applicability for beyond just OSD pods, e.g. using AWS EBS or Google PD as the underlying storage for other backends.

**Comment by yanniszark** [comment]
I think this is a necessary feature for supporting other Storage Providers seamlessly and benefiting from type reuse. 
**Assumption:** Many storage providers, including Cassandra, will choose to leverage `StatefulSets` to build their operator, as it is the de-facto way of running stateful apps on Kubernetes and also provides plenty of useful functionality out-of-the-box. 

StatefulSets use a field called `volumeClaimTemplates`, which is an array of `PersistentVolumeClaim`. The `rook/v1alpha2` common types currently lack the expressiveness to describe such a field. As a result storage providers will use their own fields (see [cockroachDB volumeSize field](https://github.com/rook/rook/blob/558d51b9ccc3cc658a8c53ff48829d9385bd59a3/pkg/apis/cockroachdb.rook.io/v1alpha1/types.go#L53)). The  `rook/v1alpha2` types should capture this commonality.

I also propose a field called `volumeClaims` or `volumeClaimTemplates` under `storageScopeSpec`, which will contain an array of `PersistentVolumeClaim`. Developers can either use it to create specific PVCs or use it as a PVC template for a StatefulSet.

**Comment by galexrt** [comment]
@yanniszark I had a look at this now and will work on this.
Exactly as you wrote I began modifying the Rook StorageSpec added a `volumeClaimTemplates` list, now I just have to add logic for the operators/controllers to consume these PVCs in addition to `devices` and `directories`.
From my knowledge of the code that should not be too hard to do as most code parts, just need a `for` loop added to them. :slightly_smiling_face:

I'll push a PR soon for feedback.

**Comment by liewegas** [comment]
One quick note:  the two primary use-case I see are (1) provisioning EBS volumes to back OSDs, and (2) provisioning NVMeoF volumes to back OSDs.  In both cases, I assume the block PVs are what we want.  File-based PVs to back OSDs are IMO of limited interest/usefulness.

**Comment by jbw976** [comment]
For 0.9, I would love to see at least the `rook.io` common types get this addition to allow specifying volume claim templates.  The full support for Ceph using the PV/PVCs does not need to be in 0.9. Thanks @galexrt!

**Comment by immesys** [comment]
If there is a way for me to help out, please mention me. I am not familiar with the codebase, but I am willing to contribute time to testing etc.

**Comment by galexrt** [comment]
@immesys Awesome! I should hopefully have some time to work on the Ceph part finally again on Friday/weekend.
I'll ping you here/in the PR, when I need your help testing it. Thanks!

**Comment by leseb** [comment]
@galexrt any updates on this? Thanks :)

**Comment by galexrt** [comment]
@leseb Sadly not. I had a working version but due to many changes to the OSD code, I haven't got it working anymore and other topics came up.

You can pick it up if you want @leseb.

***

**Quick summary**:

Most users see two points in this issue:
1. Allowing to use PVCs for `dataDirHostPath`
2. Allowing to use a list of PVCs for OSDs, cluster wide and per node.

**Comment by whereisaaron** [comment]
Being able to use PVs for Ceph is the feature I am waiting on. It is the missing link to make Rook/Ceph to fully support k8s.

Right now took/ceph relies on directly referencing devices external to k8s. This turns k8s nodes into pets, as although nodes are redundant, they aren’t cattle. If you delete a node, k8s can’t just reschedule the OSD and mount the PV on a new node. Manual intervention is required.

E.g. if you mount disks on GKE and provision rook/ceph, then next time GKE upgrades the cluster, your ceph is toast. Even though only one node is upgraded at a time, k8s has no ability to remount the ceph disk after, because ceph is directly using devices external to k8s, instead of using PVs.

If Rook/Ceph supported PVs it would be lot easier to provision and manage.

**Comment by immesys** [comment]
@whereisaaron is exactly right. There are two situations that are very difficult to handle right now. The first is the rolling cluster upgrade where nodes get killed and replaced with a new version. The second is making and restoring backups where you need to manually snapshot and restore/reattach external volumes because solutions like Velero won't do that for you. 

Both would be a ton easier if we could just use PVs

**Comment by mmgaggle** [comment]
It's also worth noting that if a PV moves to another cloud instance and the OSD pod starts there, it'll report the new host name as its parent bucket in the crush topography and this will trigger data rebalancing. While crush will try to minimize data movement, it's still going to happen unnecessarily if someone is using topologyAware: true in conjunction with a pool that uses osd-failure-domian: zone.

In this case it might be interesting to have an option to exclude a bucket type from crush the topology code. For example instead of:

```
sh-4.2# ceph osd tree
ID  CLASS WEIGHT  TYPE NAME                        STATUS REWEIGHT PRI-AFF
 -1       2.92676 root default
 -5       2.92676     region us-west-2
 -4       0.97559         zone us-west-2a
 -3       0.97559             host ip-10-0-140-53
  0   ssd 0.97559                 osd.0                up  1.00000 1.00000
-14       0.97559         zone us-west-2b
-13       0.97559             host ip-10-0-152-238
  1   ssd 0.97559                 osd.1                up  1.00000 1.00000
-10       0.97559         zone us-west-2c
 -9       0.97559             host ip-10-0-167-50
  2   ssd 0.97559                 osd.2                up  1.00000 1.00000
```

Have something like:
```
sh-4.2# ceph osd tree
ID  CLASS WEIGHT  TYPE NAME                        STATUS REWEIGHT PRI-AFF
 -1       2.92676 root default
 -5       2.92676     region us-west-2
 -4       0.97559         zone us-west-2a
  0   ssd 0.97559                 osd.0                up  1.00000 1.00000
-14       0.97559         zone us-west-2b
  1   ssd 0.97559                 osd.1                up  1.00000 1.00000
-10       0.97559         zone us-west-2c
  2   ssd 0.97559                 osd.2                up  1.00000 1.00000
```

This would result in a stable crush map despite the pod moving to a new host within the same AZ. @liewegas do you know if not having a host bucket is problematic?

While I don't see why someone would use Google Regional Persistent Disk, we may want to have scheduling logic to ensure PVs of that type do not allow the pod to be scheduled to a different zone.

**Comment by timstoop** [comment]
:tada:

**Comment by andreamlin** [comment]
@travisn Is there documentation on how to configure the rook-ceph cluster to use local PVs instead of dataDirHostPath?

**Comment by travisn** [comment]
> @travisn Is there documentation on how to configure the rook-ceph cluster to use local PVs instead of dataDirHostPath?

Yes, see the [PVC-based cluster](https://rook.github.io/docs/rook/v1.7/ceph-cluster-crd.html#pvc-based-cluster) and the [cluster-on-pvc.yaml](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/cluster-on-pvc.yaml) example.

**Comment by andreamlin** [comment]
I'm not sure I follow. Both of those examples still reference `dataDirHostPath`, and there appear to be non-trivial, non-log contents under the hostPath dir if I look at the node's filesystem. Pointing the `mon` config to a PV storageclass does seem to use a PV for the mon, but it doesn't prevent the host path usage.

**Comment by andreamlin** [comment]
If I try to remove `dataDirHostPath` from the cluster yaml, the mon pods will fail because they are still trying to write to host path at `, and we have a read-only root dir on the nodes:

```
$ kubectl -n rook-ceph describe pod rook-ceph-mon-a-6879fbf645-kc92c
...
Volumes:
  ...
  rook-ceph-log:
    Type:          HostPath (bare host directory volume)
    Path:          rook-ceph/log
...
  Warning  Failed          13s                kubelet            Error: failed to generate container "15d1a7c1c4693a36e8043c198e77193a5844862bb5194aa17887729ea233000f" spec: failed to generate spec: failed to mkdir "rook-ceph/log": mkdir rook-ceph: read-only file system
```

**Comment by travisn** [comment]
Ah, there are still some non-critical logs and crash dumps that will be written to the host path. This is still there to ensure troubleshooting can still be done on the cluster even if the pods die. We do have #5006 tracking to put even those logs in PVs, but hasn't been implemented yet. Is this a blocker for your deployment?

**Comment by andreamlin** [comment]
Thanks for the info! Not a blocker right now, but we may have security requirements that would not mesh well with hostPath in the future. I'll just keep an eye on the tracking issue 👍
